Using JSON string and HazelcastJsonValue together does not work in a query

<!--
Thanks for reporting your issue. Please share with us the following information, to help us resolve your issue quickly and efficiently.
-->

Tested on:
 
Ubuntu 20.04 on commit 725d9c5292bd969c80b503607c7459fe6929071a using java 1.8.

**Describe the bug**
The following SQL string does not work.

`client.getSql().execute("INSERT INTO test VALUES (2, CAST('[4,5,6]' AS JSON)), (3, ?)", new HazelcastJsonValue("[7,8,9]"))` throws error: 

`com.hazelcast.sql.HazelcastSqlException: java.lang.AssertionError: No assign rules for OTHER defined`

**Expected behavior**

`INSERT INTO test VALUES (2, CAST('[4,5,6]' AS JSON))"` and `INSERT INTO test VALUES (3, ?)"` with the parameter `new HazelcastJsonValue("[7,8,9]")` works. I expect the mixed usage to work, too

**To Reproduce**

Change test_client_useJsonValue test to be the following and run it:

```java
public void test_client_useJsonValue() {
        createMapping(client(), "test", "bigint", "json");

        final IMap<Long, HazelcastJsonValue> test = client().getMap("test");
        test.put(1L, json("[1,2,3]"));

        executeClient("INSERT INTO test VALUES (2, CAST('[4,5,6]' AS JSON)), (3, ?)", json("[7,8,9]"));
        assertRowsAnyOrder(client(),
                "SELECT * FROM test" ,
                rows(2,
                        1L, json("[1,2,3]"),
                        2L, json("[4,5,6]"),
                        3L, json("[7,8,9]")
                ));
    }
```


**Additional context**

```
Output: /usr/lib/jvm/java-8-openjdk-amd64/bin/java -ea -Xms128m -Xmx1G -Dhazelcast.phone.home.enabled=false -Dhazelcast.test.use.network=false -Dlog4j.skipJansi=true -Xms128m -Xmx1G -Dhazelcast.phone.home.enabled=false -Dhazelcast.test.use.network=false -Dlog4j.skipJansi=true -Didea.test.cyclic.buffer.size=1048576 -javaagent:/home/serkan/idea-IU-213.5744.223/lib/idea_rt.jar=39277:/home/serkan/idea-IU-213.5744.223/bin -Dfile.encoding=UTF-8 -classpath /home/serkan/idea-IU-213.5744.223/lib/idea_rt.jar:/home/serkan/idea-IU-213.5744.223/plugins/junit/lib/junit5-rt.jar:/home/serkan/idea-IU-213.5744.223/plugins/junit/lib/junit-rt.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/charsets.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/cldrdata.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/dnsns.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/icedtea-sound.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/jaccess.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/java-atk-wrapper.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/localedata.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/nashorn.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/sunec.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/sunjce_provider.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/sunpkcs11.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/ext/zipfs.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jce.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jfr.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jsse.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/management-agent.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/resources.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar:/home/serkan/forked/hazelcast/hazelcast-sql/target/test-classes:/home/serkan/forked/hazelcast/hazelcast-sql/target/classes:/home/serkan/forked/hazelcast/hazelcast/target/classes:/home/serkan/.m2/repository/com/fasterxml/jackson/jr/jackson-jr-objects/2.12.1/jackson-jr-objects-2.12.1.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/jr/jackson-jr-annotation-support/2.12.1/jackson-jr-annotation-support-2.12.1.jar:/home/serkan/.m2/repository/org/objenesis/objenesis/3.1/objenesis-3.1.jar:/home/serkan/.m2/repository/io/github/classgraph/classgraph/4.8.66/classgraph-4.8.66.jar:/home/serkan/.m2/repository/com/github/erosb/everit-json-schema/1.13.0/everit-json-schema-1.13.0.jar:/home/serkan/.m2/repository/org/json/json/20201115/json-20201115.jar:/home/serkan/.m2/repository/org/apache/calcite/calcite-core/1.28.0/calcite-core-1.28.0.jar:/home/serkan/.m2/repository/org/apache/calcite/calcite-linq4j/1.28.0/calcite-linq4j-1.28.0.jar:/home/serkan/.m2/repository/com/esri/geometry/esri-geometry-api/2.2.0/esri-geometry-api-2.2.0.jar:/home/serkan/.m2/repository/org/apache/calcite/avatica/avatica-core/1.19.0/avatica-core-1.19.0.jar:/home/serkan/.m2/repository/org/checkerframework/checker-qual/3.10.0/checker-qual-3.10.0.jar:/home/serkan/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib-jdk8/1.5.31/kotlin-stdlib-jdk8-1.5.31.jar:/home/serkan/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib/1.5.31/kotlin-stdlib-1.5.31.jar:/home/serkan/.m2/repository/org/jetbrains/annotations/13.0/annotations-13.0.jar:/home/serkan/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib-common/1.5.31/kotlin-stdlib-common-1.5.31.jar:/home/serkan/.m2/repository/org/jetbrains/kotlin/kotlin-stdlib-jdk7/1.5.31/kotlin-stdlib-jdk7-1.5.31.jar:/home/serkan/.m2/repository/com/google/uzaygezen/uzaygezen-core/0.2/uzaygezen-core-0.2.jar:/home/serkan/.m2/repository/commons-codec/commons-codec/1.15/commons-codec-1.15.jar:/home/serkan/.m2/repository/org/codehaus/janino/commons-compiler/3.1.6/commons-compiler-3.1.6.jar:/home/serkan/.m2/repository/org/codehaus/janino/janino/3.1.6/janino-3.1.6.jar:/home/serkan/.m2/repository/com/google/guava/guava/30.1.1-jre/guava-30.1.1-jre.jar:/home/serkan/.m2/repository/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar:/home/serkan/.m2/repository/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar:/home/serkan/.m2/repository/org/slf4j/slf4j-api/1.7.30/slf4j-api-1.7.30.jar:/home/serkan/.m2/repository/com/jayway/jsonpath/json-path/2.4.0/json-path-2.4.0.jar:/home/serkan/.m2/repository/net/minidev/json-smart/2.4.7/json-smart-2.4.7.jar:/home/serkan/.m2/repository/net/minidev/accessors-smart/2.4.7/accessors-smart-2.4.7.jar:/home/serkan/.m2/repository/org/ow2/asm/asm/9.1/asm-9.1.jar:/home/serkan/.m2/repository/com/google/code/gson/gson/2.8.6/gson-2.8.6.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.12.1/jackson-core-2.12.1.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.12.1/jackson-databind-2.12.1.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.12.1/jackson-annotations-2.12.1.jar:/home/serkan/.m2/repository/org/apache/avro/avro/1.11.0/avro-1.11.0.jar:/home/serkan/.m2/repository/org/apache/commons/commons-compress/1.21/commons-compress-1.21.jar:/home/serkan/forked/hazelcast/extensions/kafka/target/classes:/home/serkan/.m2/repository/org/apache/kafka/kafka-clients/2.2.2/kafka-clients-2.2.2.jar:/home/serkan/.m2/repository/com/github/luben/zstd-jni/1.4.5-12/zstd-jni-1.4.5-12.jar:/home/serkan/.m2/repository/org/lz4/lz4-java/1.5.0/lz4-java-1.5.0.jar:/home/serkan/.m2/repository/org/xerial/snappy/snappy-java/1.1.7.2/snappy-java-1.1.7.2.jar:/home/serkan/forked/hazelcast/hazelcast/target/test-classes:/home/serkan/.m2/repository/pl/pragmatists/JUnitParams/1.1.1/JUnitParams-1.1.1.jar:/home/serkan/forked/hazelcast/extensions/avro/target/classes:/home/serkan/forked/hazelcast/extensions/csv/target/classes:/home/serkan/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-csv/2.12.1/jackson-dataformat-csv-2.12.1.jar:/home/serkan/forked/hazelcast/extensions/hadoop/target/classes:/home/serkan/.m2/repository/org/apache/avro/avro-mapred/1.11.0/avro-mapred-1.11.0.jar:/home/serkan/.m2/repository/org/apache/parquet/parquet-avro/1.12.0/parquet-avro-1.12.0.jar:/home/serkan/.m2/repository/org/apache/parquet/parquet-column/1.12.0/parquet-column-1.12.0.jar:/home/serkan/.m2/repository/org/apache/parquet/parquet-common/1.12.0/parquet-common-1.12.0.jar:/home/serkan/.m2/repository/org/apache/parquet/parquet-encoding/1.12.0/parquet-encoding-1.12.0.jar:/home/serkan/.m2/repository/org/apache/parquet/parquet-hadoop/1.12.0/parquet-hadoop-1.12.0.jar:/home/serkan/.m2/repository/org/apache/parquet/parquet-jackson/1.12.0/parquet-jackson-1.12.0.jar:/home/serkan/.m2/repository/commons-pool/commons-pool/1.6/commons-pool-1.6.jar:/home/serkan/.m2/repository/org/apache/parquet/parquet-format-structures/1.12.0/parquet-format-structures-1.12.0.jar:/home/serkan/.m2/repository/javax/annotation/javax.annotation-api/1.3.2/javax.annotation-api-1.3.2.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-minicluster/3.3.1/hadoop-minicluster-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-common/3.3.1/hadoop-common-3.3.1-tests.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-hdfs/3.3.1/hadoop-hdfs-3.3.1-tests.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-util-ajax/9.4.43.v20210629/jetty-util-ajax-9.4.43.v20210629.jar:/home/serkan/.m2/repository/commons-daemon/commons-daemon/1.0.13/commons-daemon-1.0.13.jar:/home/serkan/.m2/repository/io/netty/netty/3.10.6.Final/netty-3.10.6.Final.jar:/home/serkan/.m2/repository/io/netty/netty-all/4.1.63.Final/netty-all-4.1.63.Final.jar:/home/serkan/.m2/repository/org/fusesource/leveldbjni/leveldbjni-all/1.8/leveldbjni-all-1.8.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-server-tests/3.3.1/hadoop-yarn-server-tests-3.3.1-tests.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-server-common/3.3.1/hadoop-yarn-server-common-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-registry/3.3.1/hadoop-registry-3.3.1.jar:/home/serkan/.m2/repository/org/apache/geronimo/specs/geronimo-jcache_1.0_spec/1.0-alpha-1/geronimo-jcache_1.0_spec-1.0-alpha-1.jar:/home/serkan/.m2/repository/org/ehcache/ehcache/3.3.1/ehcache-3.3.1.jar:/home/serkan/.m2/repository/com/zaxxer/HikariCP-java7/2.4.12/HikariCP-java7-2.4.12.jar:/home/serkan/.m2/repository/com/microsoft/sqlserver/mssql-jdbc/6.2.1.jre7/mssql-jdbc-6.2.1.jre7.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-server-nodemanager/3.3.1/hadoop-yarn-server-nodemanager-3.3.1.jar:/home/serkan/.m2/repository/com/sun/jersey/jersey-client/1.19/jersey-client-1.19.jar:/home/serkan/.m2/repository/org/eclipse/jetty/websocket/javax-websocket-server-impl/9.4.43.v20210629/javax-websocket-server-impl-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-annotations/9.4.43.v20210629/jetty-annotations-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-plus/9.4.43.v20210629/jetty-plus-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-jndi/9.4.43.v20210629/jetty-jndi-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/ow2/asm/asm-commons/9.1/asm-commons-9.1.jar:/home/serkan/.m2/repository/org/ow2/asm/asm-tree/9.1/asm-tree-9.1.jar:/home/serkan/.m2/repository/org/ow2/asm/asm-analysis/9.1/asm-analysis-9.1.jar:/home/serkan/.m2/repository/org/eclipse/jetty/websocket/javax-websocket-client-impl/9.4.43.v20210629/javax-websocket-client-impl-9.4.43.v20210629.jar:/home/serkan/.m2/repository/javax/websocket/javax.websocket-client-api/1.0/javax.websocket-client-api-1.0.jar:/home/serkan/.m2/repository/org/eclipse/jetty/websocket/websocket-server/9.4.43.v20210629/websocket-server-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/websocket/websocket-servlet/9.4.43.v20210629/websocket-servlet-9.4.43.v20210629.jar:/home/serkan/.m2/repository/javax/websocket/javax.websocket-api/1.0/javax.websocket-api-1.0.jar:/home/serkan/.m2/repository/io/dropwizard/metrics/metrics-core/3.2.4/metrics-core-3.2.4.jar:/home/serkan/.m2/repository/com/google/inject/guice/4.0/guice-4.0.jar:/home/serkan/.m2/repository/javax/inject/javax.inject/1/javax.inject-1.jar:/home/serkan/.m2/repository/aopalliance/aopalliance/1.0/aopalliance-1.0.jar:/home/serkan/.m2/repository/com/sun/jersey/contribs/jersey-guice/1.19/jersey-guice-1.19.jar:/home/serkan/.m2/repository/net/java/dev/jna/jna/5.2.0/jna-5.2.0.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-server-resourcemanager/3.3.1/hadoop-yarn-server-resourcemanager-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-server-applicationhistoryservice/3.3.1/hadoop-yarn-server-applicationhistoryservice-3.3.1.jar:/home/serkan/.m2/repository/de/ruedigermoeller/fst/2.50/fst-2.50.jar:/home/serkan/.m2/repository/com/cedarsoftware/java-util/1.9.0/java-util-1.9.0.jar:/home/serkan/.m2/repository/com/cedarsoftware/json-io/2.5.1/json-io-2.5.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-server-timelineservice/3.3.1/hadoop-yarn-server-timelineservice-3.3.1.jar:/home/serkan/.m2/repository/org/apache/commons/commons-csv/1.0/commons-csv-1.0.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-common/3.3.1/hadoop-yarn-common-3.3.1.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/module/jackson-module-jaxb-annotations/2.12.1/jackson-module-jaxb-annotations-2.12.1.jar:/home/serkan/.m2/repository/jakarta/xml/bind/jakarta.xml.bind-api/2.3.2/jakarta.xml.bind-api-2.3.2.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-mapreduce-client-jobclient/3.3.1/hadoop-mapreduce-client-jobclient-3.3.1-tests.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-mapreduce-client-common/3.3.1/hadoop-mapreduce-client-common-3.3.1.jar:/home/serkan/.m2/repository/com/google/inject/extensions/guice-servlet/4.0/guice-servlet-4.0.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-hdfs/3.3.1/hadoop-hdfs-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-mapreduce-client-app/3.3.1/hadoop-mapreduce-client-app-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-mapreduce-client-shuffle/3.3.1/hadoop-mapreduce-client-shuffle-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-api/3.3.1/hadoop-yarn-api-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-mapreduce-client-core/3.3.1/hadoop-mapreduce-client-core-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-client/3.3.1/hadoop-yarn-client-3.3.1.jar:/home/serkan/.m2/repository/org/eclipse/jetty/websocket/websocket-client/9.4.43.v20210629/websocket-client-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-client/9.4.43.v20210629/jetty-client-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/websocket/websocket-common/9.4.43.v20210629/websocket-common-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/websocket/websocket-api/9.4.43.v20210629/websocket-api-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/jline/jline/3.19.0/jline-3.19.0.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-hdfs-client/3.3.1/hadoop-hdfs-client-3.3.1.jar:/home/serkan/.m2/repository/com/squareup/okhttp/okhttp/2.7.5/okhttp-2.7.5.jar:/home/serkan/.m2/repository/com/squareup/okio/okio/1.6.0/okio-1.6.0.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-mapreduce-client-jobclient/3.3.1/hadoop-mapreduce-client-jobclient-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-mapreduce-client-hs/3.3.1/hadoop-mapreduce-client-hs-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-yarn-server-web-proxy/3.3.1/hadoop-yarn-server-web-proxy-3.3.1.jar:/home/serkan/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.60/bcprov-jdk15on-1.60.jar:/home/serkan/.m2/repository/org/bouncycastle/bcpkix-jdk15on/1.60/bcpkix-jdk15on-1.60.jar:/home/serkan/.m2/repository/org/apache/kafka/kafka_2.12/2.2.2/kafka_2.12-2.2.2.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/datatype/jackson-datatype-jdk8/2.12.1/jackson-datatype-jdk8-2.12.1.jar:/home/serkan/.m2/repository/net/sf/jopt-simple/jopt-simple/5.0.4/jopt-simple-5.0.4.jar:/home/serkan/.m2/repository/com/yammer/metrics/metrics-core/2.2.0/metrics-core-2.2.0.jar:/home/serkan/.m2/repository/org/scala-lang/scala-library/2.12.8/scala-library-2.12.8.jar:/home/serkan/.m2/repository/org/scala-lang/scala-reflect/2.12.8/scala-reflect-2.12.8.jar:/home/serkan/.m2/repository/com/typesafe/scala-logging/scala-logging_2.12/3.9.0/scala-logging_2.12-3.9.0.jar:/home/serkan/.m2/repository/com/101tec/zkclient/0.11/zkclient-0.11.jar:/home/serkan/.m2/repository/org/apache/zookeeper/zookeeper/3.5.8/zookeeper-3.5.8.jar:/home/serkan/.m2/repository/org/apache/zookeeper/zookeeper-jute/3.5.8/zookeeper-jute-3.5.8.jar:/home/serkan/.m2/repository/org/apache/yetus/audience-annotations/0.12.0/audience-annotations-0.12.0.jar:/home/serkan/.m2/repository/io/netty/netty-handler/4.1.63.Final/netty-handler-4.1.63.Final.jar:/home/serkan/.m2/repository/io/netty/netty-common/4.1.63.Final/netty-common-4.1.63.Final.jar:/home/serkan/.m2/repository/io/netty/netty-resolver/4.1.63.Final/netty-resolver-4.1.63.Final.jar:/home/serkan/.m2/repository/io/netty/netty-buffer/4.1.63.Final/netty-buffer-4.1.63.Final.jar:/home/serkan/.m2/repository/io/netty/netty-transport/4.1.63.Final/netty-transport-4.1.63.Final.jar:/home/serkan/.m2/repository/io/netty/netty-codec/4.1.63.Final/netty-codec-4.1.63.Final.jar:/home/serkan/.m2/repository/io/netty/netty-transport-native-epoll/4.1.63.Final/netty-transport-native-epoll-4.1.63.Final.jar:/home/serkan/.m2/repository/io/netty/netty-transport-native-unix-common/4.1.63.Final/netty-transport-native-unix-common-4.1.63.Final.jar:/home/serkan/.m2/repository/log4j/log4j/1.2.17/log4j-1.2.17.jar:/home/serkan/.m2/repository/io/confluent/kafka-avro-serializer/4.1.4/kafka-avro-serializer-4.1.4.jar:/home/serkan/.m2/repository/io/confluent/kafka-schema-registry-client/4.1.4/kafka-schema-registry-client-4.1.4.jar:/home/serkan/.m2/repository/io/confluent/common-config/4.1.4/common-config-4.1.4.jar:/home/serkan/.m2/repository/io/confluent/common-utils/4.1.4/common-utils-4.1.4.jar:/home/serkan/.m2/repository/io/confluent/kafka-schema-registry/4.1.4/kafka-schema-registry-4.1.4.jar:/home/serkan/.m2/repository/org/apache/kafka/kafka_2.11/1.1.1-cp4/kafka_2.11-1.1.1-cp4.jar:/home/serkan/.m2/repository/com/typesafe/scala-logging/scala-logging_2.11/3.8.0/scala-logging_2.11-3.8.0.jar:/home/serkan/.m2/repository/org/glassfish/jersey/ext/jersey-bean-validation/2.25/jersey-bean-validation-2.25.jar:/home/serkan/.m2/repository/org/glassfish/hk2/external/javax.inject/2.5.0-b30/javax.inject-2.5.0-b30.jar:/home/serkan/.m2/repository/org/glassfish/jersey/core/jersey-common/2.25/jersey-common-2.25.jar:/home/serkan/.m2/repository/org/glassfish/jersey/bundles/repackaged/jersey-guava/2.25/jersey-guava-2.25.jar:/home/serkan/.m2/repository/org/glassfish/hk2/hk2-api/2.5.0-b30/hk2-api-2.5.0-b30.jar:/home/serkan/.m2/repository/org/glassfish/hk2/hk2-utils/2.5.0-b30/hk2-utils-2.5.0-b30.jar:/home/serkan/.m2/repository/org/glassfish/hk2/external/aopalliance-repackaged/2.5.0-b30/aopalliance-repackaged-2.5.0-b30.jar:/home/serkan/.m2/repository/org/glassfish/hk2/hk2-locator/2.5.0-b30/hk2-locator-2.5.0-b30.jar:/home/serkan/.m2/repository/org/glassfish/hk2/osgi-resource-locator/1.0.1/osgi-resource-locator-1.0.1.jar:/home/serkan/.m2/repository/org/glassfish/jersey/core/jersey-server/2.25/jersey-server-2.25.jar:/home/serkan/.m2/repository/org/glassfish/jersey/core/jersey-client/2.25/jersey-client-2.25.jar:/home/serkan/.m2/repository/org/glassfish/jersey/media/jersey-media-jaxb/2.25/jersey-media-jaxb-2.25.jar:/home/serkan/.m2/repository/javax/validation/validation-api/1.1.0.Final/validation-api-1.1.0.Final.jar:/home/serkan/.m2/repository/org/hibernate/hibernate-validator/5.1.3.Final/hibernate-validator-5.1.3.Final.jar:/home/serkan/.m2/repository/org/jboss/logging/jboss-logging/3.1.3.GA/jboss-logging-3.1.3.GA.jar:/home/serkan/.m2/repository/com/fasterxml/classmate/1.0.0/classmate-1.0.0.jar:/home/serkan/.m2/repository/javax/el/javax.el-api/2.2.4/javax.el-api-2.2.4.jar:/home/serkan/.m2/repository/org/glassfish/web/javax.el/2.2.4/javax.el-2.2.4.jar:/home/serkan/.m2/repository/javax/ws/rs/javax.ws.rs-api/2.0.1/javax.ws.rs-api-2.0.1.jar:/home/serkan/.m2/repository/io/confluent/rest-utils/4.1.4/rest-utils-4.1.4.jar:/home/serkan/.m2/repository/io/confluent/common-metrics/4.1.4/common-metrics-4.1.4.jar:/home/serkan/.m2/repository/org/glassfish/jersey/containers/jersey-container-servlet/2.25/jersey-container-servlet-2.25.jar:/home/serkan/.m2/repository/org/glassfish/jersey/containers/jersey-container-servlet-core/2.25/jersey-container-servlet-core-2.25.jar:/home/serkan/.m2/repository/javax/activation/activation/1.1.1/activation-1.1.1.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-jmx/9.4.43.v20210629/jetty-jmx-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-servlets/9.4.43.v20210629/jetty-servlets-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-continuation/9.4.43.v20210629/jetty-continuation-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-jaas/9.4.43.v20210629/jetty-jaas-9.4.43.v20210629.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/jaxrs/jackson-jaxrs-json-provider/2.12.1/jackson-jaxrs-json-provider-2.12.1.jar:/home/serkan/.m2/repository/com/fasterxml/jackson/jaxrs/jackson-jaxrs-base/2.12.1/jackson-jaxrs-base-2.12.1.jar:/home/serkan/.m2/repository/javax/xml/bind/jaxb-api/2.3.1/jaxb-api-2.3.1.jar:/home/serkan/.m2/repository/javax/activation/javax.activation-api/1.2.0/javax.activation-api-1.2.0.jar:/home/serkan/forked/hazelcast/extensions/kafka/target/test-classes:/home/serkan/.m2/repository/org/apache/kafka/kafka_2.12/2.2.2/kafka_2.12-2.2.2-test.jar:/home/serkan/.m2/repository/org/apache/kafka/kafka-clients/2.2.2/kafka-clients-2.2.2-test.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-common/3.3.1/hadoop-common-3.3.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/thirdparty/hadoop-shaded-protobuf_3_7/1.1.1/hadoop-shaded-protobuf_3_7-1.1.1.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-annotations/3.3.1/hadoop-annotations-3.3.1.jar:/usr/lib/jvm/java-8-openjdk-amd64/lib/tools.jar:/home/serkan/.m2/repository/org/apache/hadoop/thirdparty/hadoop-shaded-guava/1.1.1/hadoop-shaded-guava-1.1.1.jar:/home/serkan/.m2/repository/commons-cli/commons-cli/1.2/commons-cli-1.2.jar:/home/serkan/.m2/repository/org/apache/commons/commons-math3/3.1.1/commons-math3-3.1.1.jar:/home/serkan/.m2/repository/org/apache/httpcomponents/httpclient/4.5.13/httpclient-4.5.13.jar:/home/serkan/.m2/repository/org/apache/httpcomponents/httpcore/4.4.13/httpcore-4.4.13.jar:/home/serkan/.m2/repository/commons-io/commons-io/2.8.0/commons-io-2.8.0.jar:/home/serkan/.m2/repository/commons-net/commons-net/3.6/commons-net-3.6.jar:/home/serkan/.m2/repository/commons-collections/commons-collections/3.2.2/commons-collections-3.2.2.jar:/home/serkan/.m2/repository/javax/servlet/javax.servlet-api/3.1.0/javax.servlet-api-3.1.0.jar:/home/serkan/.m2/repository/jakarta/activation/jakarta.activation-api/1.2.1/jakarta.activation-api-1.2.1.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-server/9.4.43.v20210629/jetty-server-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-http/9.4.43.v20210629/jetty-http-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-io/9.4.43.v20210629/jetty-io-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-util/9.4.43.v20210629/jetty-util-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-servlet/9.4.43.v20210629/jetty-servlet-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-security/9.4.43.v20210629/jetty-security-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-webapp/9.4.43.v20210629/jetty-webapp-9.4.43.v20210629.jar:/home/serkan/.m2/repository/org/eclipse/jetty/jetty-xml/9.4.43.v20210629/jetty-xml-9.4.43.v20210629.jar:/home/serkan/.m2/repository/javax/servlet/jsp/jsp-api/2.1/jsp-api-2.1.jar:/home/serkan/.m2/repository/com/sun/jersey/jersey-core/1.19/jersey-core-1.19.jar:/home/serkan/.m2/repository/javax/ws/rs/jsr311-api/1.1.1/jsr311-api-1.1.1.jar:/home/serkan/.m2/repository/com/sun/jersey/jersey-servlet/1.19/jersey-servlet-1.19.jar:/home/serkan/.m2/repository/com/sun/jersey/jersey-json/1.19/jersey-json-1.19.jar:/home/serkan/.m2/repository/org/codehaus/jettison/jettison/1.1/jettison-1.1.jar:/home/serkan/.m2/repository/com/sun/xml/bind/jaxb-impl/2.2.3-1/jaxb-impl-2.2.3-1.jar:/home/serkan/.m2/repository/org/codehaus/jackson/jackson-core-asl/1.9.2/jackson-core-asl-1.9.2.jar:/home/serkan/.m2/repository/org/codehaus/jackson/jackson-mapper-asl/1.9.2/jackson-mapper-asl-1.9.2.jar:/home/serkan/.m2/repository/org/codehaus/jackson/jackson-jaxrs/1.9.2/jackson-jaxrs-1.9.2.jar:/home/serkan/.m2/repository/org/codehaus/jackson/jackson-xc/1.9.2/jackson-xc-1.9.2.jar:/home/serkan/.m2/repository/com/sun/jersey/jersey-server/1.19/jersey-server-1.19.jar:/home/serkan/.m2/repository/commons-logging/commons-logging/1.2/commons-logging-1.2.jar:/home/serkan/.m2/repository/commons-beanutils/commons-beanutils/1.9.4/commons-beanutils-1.9.4.jar:/home/serkan/.m2/repository/org/apache/commons/commons-configuration2/2.1.1/commons-configuration2-2.1.1.jar:/home/serkan/.m2/repository/org/apache/commons/commons-lang3/3.11/commons-lang3-3.11.jar:/home/serkan/.m2/repository/org/apache/commons/commons-text/1.4/commons-text-1.4.jar:/home/serkan/.m2/repository/com/google/re2j/re2j/1.1/re2j-1.1.jar:/home/serkan/.m2/repository/com/google/protobuf/protobuf-java/3.13.0/protobuf-java-3.13.0.jar:/home/serkan/.m2/repository/org/apache/hadoop/hadoop-auth/3.3.1/hadoop-auth-3.3.1.jar:/home/serkan/.m2/repository/com/nimbusds/nimbus-jose-jwt/7.9/nimbus-jose-jwt-7.9.jar:/home/serkan/.m2/repository/com/github/stephenc/jcip/jcip-annotations/1.0-1/jcip-annotations-1.0-1.jar:/home/serkan/.m2/repository/org/apache/curator/curator-framework/4.2.0/curator-framework-4.2.0.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-simplekdc/1.0.1/kerb-simplekdc-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-client/1.0.1/kerb-client-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerby-config/1.0.1/kerby-config-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-common/1.0.1/kerb-common-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-crypto/1.0.1/kerb-crypto-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-util/1.0.1/kerb-util-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/token-provider/1.0.1/token-provider-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-admin/1.0.1/kerb-admin-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-server/1.0.1/kerb-server-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-identity/1.0.1/kerb-identity-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerby-xdr/1.0.1/kerby-xdr-1.0.1.jar:/home/serkan/.m2/repository/com/jcraft/jsch/0.1.55/jsch-0.1.55.jar:/home/serkan/.m2/repository/org/apache/curator/curator-client/4.2.0/curator-client-4.2.0.jar:/home/serkan/.m2/repository/org/apache/curator/curator-recipes/4.2.0/curator-recipes-4.2.0.jar:/home/serkan/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar:/home/serkan/.m2/repository/org/apache/htrace/htrace-core4/4.1.0-incubating/htrace-core4-4.1.0-incubating.jar:/home/serkan/.m2/repository/org/apache/kerby/kerb-core/1.0.1/kerb-core-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerby-pkix/1.0.1/kerby-pkix-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerby-asn1/1.0.1/kerby-asn1-1.0.1.jar:/home/serkan/.m2/repository/org/apache/kerby/kerby-util/1.0.1/kerby-util-1.0.1.jar:/home/serkan/.m2/repository/org/codehaus/woodstox/stax2-api/4.2.1/stax2-api-4.2.1.jar:/home/serkan/.m2/repository/com/fasterxml/woodstox/woodstox-core/5.3.0/woodstox-core-5.3.0.jar:/home/serkan/.m2/repository/dnsjava/dnsjava/2.1.7/dnsjava-2.1.7.jar:/home/serkan/.m2/repository/javax/cache/cache-api/1.1.1/cache-api-1.1.1.jar:/home/serkan/.m2/repository/org/snakeyaml/snakeyaml-engine/2.1/snakeyaml-engine-2.1.jar:/home/serkan/.m2/repository/junit/junit/4.13.2/junit-4.13.2.jar:/home/serkan/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar:/home/serkan/.m2/repository/org/hamcrest/hamcrest-library/1.3/hamcrest-library-1.3.jar:/home/serkan/.m2/repository/org/assertj/assertj-core/3.15.0/assertj-core-3.15.0.jar:/home/serkan/.m2/repository/org/mockito/mockito-core/3.6.0/mockito-core-3.6.0.jar:/home/serkan/.m2/repository/org/powermock/powermock-api-mockito2/2.0.9/powermock-api-mockito2-2.0.9.jar:/home/serkan/.m2/repository/org/powermock/powermock-api-support/2.0.9/powermock-api-support-2.0.9.jar:/home/serkan/.m2/repository/org/powermock/powermock-reflect/2.0.9/powermock-reflect-2.0.9.jar:/home/serkan/.m2/repository/org/powermock/powermock-core/2.0.9/powermock-core-2.0.9.jar:/home/serkan/.m2/repository/org/powermock/powermock-module-junit4/2.0.9/powermock-module-junit4-2.0.9.jar:/home/serkan/.m2/repository/org/powermock/powermock-module-junit4-common/2.0.9/powermock-module-junit4-common-2.0.9.jar:/home/serkan/.m2/repository/org/reflections/reflections/0.9.10/reflections-0.9.10.jar:/home/serkan/.m2/repository/org/javassist/javassist/3.26.0-GA/javassist-3.26.0-GA.jar:/home/serkan/.m2/repository/org/openjdk/jmh/jmh-core/1.27/jmh-core-1.27.jar:/home/serkan/.m2/repository/org/openjdk/jmh/jmh-generator-annprocess/1.27/jmh-generator-annprocess-1.27.jar:/home/serkan/.m2/repository/org/apache/logging/log4j/log4j-core/2.17.0/log4j-core-2.17.0.jar:/home/serkan/.m2/repository/org/apache/logging/log4j/log4j-api/2.17.0/log4j-api-2.17.0.jar:/home/serkan/.m2/repository/com/google/code/findbugs/annotations/3.0.1u2/annotations-3.0.1u2.jar:/home/serkan/.m2/repository/net/jcip/jcip-annotations/1.0/jcip-annotations-1.0.jar:/home/serkan/.m2/repository/net/bytebuddy/byte-buddy/1.10.21/byte-buddy-1.10.21.jar:/home/serkan/.m2/repository/net/bytebuddy/byte-buddy-agent/1.10.21/byte-buddy-agent-1.10.21.jar:/home/serkan/.m2/repository/nl/jqno/equalsverifier/equalsverifier/3.8/equalsverifier-3.8.jar:/home/serkan/.m2/repository/org/apache/tomcat/embed/tomcat-embed-core/9.0.41/tomcat-embed-core-9.0.41.jar:/home/serkan/.m2/repository/org/apache/tomcat/tomcat-annotations-api/9.0.41/tomcat-annotations-api-9.0.41.jar com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit4 com.hazelcast.jet.sql.SqlJsonTypeTest,test_client_useJsonValue
10:22:04,499  INFO || - [MetricsConfigHelper] Time-limited test - [LOCAL] [dev] [5.1-SNAPSHOT] Overridden metrics configuration with system property 'hazelcast.metrics.collection.frequency'='1' -> 'MetricsConfig.collectionFrequencySeconds'='1'
10:22:04,708  INFO || - [logo] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] 
	+       +  o    o     o     o---o o----o o      o---o     o     o----o o--o--o
	+ +   + +  |    |    / \       /  |      |     /         / \    |         |   
	+ + + + +  o----o   o   o     o   o----o |    o         o   o   o----o    |   
	+ +   + +  |    |  /     \   /    |      |     \       /     \       |    |   
	+       +  o    o o       o o---o o----o o----o o---o o       o o----o    o   
10:22:04,709  INFO || - [system] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
10:22:04,709  INFO || - [system] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Hazelcast Platform 5.1-SNAPSHOT (20220104 - 725d9c5) starting at [127.0.0.1]:5701
10:22:04,710  INFO || - [system] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Cluster name: dev
10:22:05,492  INFO || - [MetricsConfigHelper] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Collecting debug metrics and sending to diagnostics is enabled
10:22:06,536  WARN || - [CPSubsystem] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
10:22:07,324  INFO || - [JetServiceBackend] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Setting number of cooperative threads and default parallelism to 2
10:22:07,579  INFO || - [Diagnostics] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
10:22:07,614  INFO || - [LifecycleService] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] [127.0.0.1]:5701 is STARTING
10:22:07,647  INFO || - [JetExtension] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Jet is enabled
10:22:07,649  INFO || - [ClusterService] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5701 - b87c91fa-10ad-49a3-8c23-2840f726f150 this
]

10:22:07,649 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:07,676  INFO || - [LifecycleService] Time-limited test - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] [127.0.0.1]:5701 is STARTED
10:22:07,750 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:07,800  INFO || - [ClientInvocationService] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] Running with 2 response threads, dynamic=true
10:22:07,843  INFO || - [LifecycleService] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] HazelcastClient 5.1-SNAPSHOT (20220104 - 725d9c5) is STARTING
10:22:07,845  INFO || - [LifecycleService] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] HazelcastClient 5.1-SNAPSHOT (20220104 - 725d9c5) is STARTED
10:22:07,851 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:07,866  INFO || - [ClientConnectionManager] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] Trying to connect to cluster: dev
10:22:07,872  INFO || - [ClientConnectionManager] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] Trying to connect to [127.0.0.1]:5701
10:22:07,899  INFO || - [HazelcastClient] Time-limited test - Created connection to endpoint: [127.0.0.1]:5701, connection: MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=null, lastReadTime=never, lastWriteTime=never, closedTime=never, connected server version=null}}
10:22:07,934 DEBUG || - [AuthenticationMessageTask] hz.happy_perlman.priority-generic-operation.thread-0 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Processing authentication with clientUuid fd4d94fe-7b69-47d7-a12d-6dcfe7adfd7f and clientName hz.client_1
10:22:07,939  INFO || - [AuthenticationMessageTask] hz.happy_perlman.priority-generic-operation.thread-0 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Received auth from MockedNodeConnection{ remoteAddress = [127.0.0.1]:40001, localAddress = [127.0.0.1]:5701, connectionId = 1}, successfully authenticated, clientUuid: fd4d94fe-7b69-47d7-a12d-6dcfe7adfd7f, client name: hz.client_1, client version: 5.1-SNAPSHOT
10:22:07,952 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:07,955  INFO || - [LifecycleService] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] HazelcastClient 5.1-SNAPSHOT (20220104 - 725d9c5) is CLIENT_CONNECTED
10:22:07,955  INFO || - [ClientConnectionManager] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] Authenticated with server [127.0.0.1]:5701:b87c91fa-10ad-49a3-8c23-2840f726f150, server version: 5.1-SNAPSHOT, local address: /127.0.0.1:40001
10:22:07,959  INFO || - [Diagnostics] Time-limited test - hz.client_1 [dev] [5.1-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
10:22:07,969 TRACE || - [ClientListenerService] hz.client_1.internal-1 - hz.client_1 [dev] [5.1-SNAPSHOT] Register attempt of ClusterViewListenerHandler to MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-01-04 10:22:07.943, lastWriteTime=2022-01-04 10:22:07.919, closedTime=never, connected server version=5.1-SNAPSHOT}}
10:22:07,987  INFO || - [ClientClusterService] hz.client_1.event-5 - hz.client_1 [dev] [5.1-SNAPSHOT] 

Members [1] {
	Member [127.0.0.1]:5701 - b87c91fa-10ad-49a3-8c23-2840f726f150
}

10:22:07,996 TRACE || - [ClientListenerService] ForkJoinPool.commonPool-worker-9 - hz.client_1 [dev] [5.1-SNAPSHOT] Registered ClusterViewListenerHandler to MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-01-04 10:22:07.986, lastWriteTime=2022-01-04 10:22:07.970, closedTime=never, connected server version=5.1-SNAPSHOT}}
10:22:08,050 TRACE || - [ClientListenerService] hz.client_1.eventRegistration- - hz.client_1 [dev] [5.1-SNAPSHOT] Register attempt of ClientListenerRegistration{codec=com.hazelcast.client.impl.spi.impl.ClientInvocationServiceImpl$1@74149e5e, handler=com.hazelcast.client.impl.spi.impl.ClientInvocationServiceImpl$BackupEventHandler@7ad36eac} to MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-01-04 10:22:07.986, lastWriteTime=2022-01-04 10:22:07.970, closedTime=never, connected server version=5.1-SNAPSHOT}}
10:22:08,053 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:08,059 TRACE || - [ClientListenerService] hz.client_1.eventRegistration- - hz.client_1 [dev] [5.1-SNAPSHOT] Registered ClientListenerRegistration{codec=com.hazelcast.client.impl.spi.impl.ClientInvocationServiceImpl$1@74149e5e, handler=com.hazelcast.client.impl.spi.impl.ClientInvocationServiceImpl$BackupEventHandler@7ad36eac} to MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=true, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-01-04 10:22:08.057, lastWriteTime=2022-01-04 10:22:08.050, closedTime=never, connected server version=5.1-SNAPSHOT}}
10:22:08,071  INFO || - [ClientStatisticsService] Time-limited test - Client statistics is enabled with period 5 seconds.
Started Running Test: test_client_useJsonValue
10:22:08,154 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:08,401 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:08,503 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:08,604 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:08,704 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:08,805 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:08,906 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,007 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,108 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,209 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,309 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,410 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,511 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,612 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,712 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,813 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:09,914 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:10,015 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:10,116 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:10,216 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:10,317 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:10,418 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:10,518 DEBUG || - [JobCoordinationService] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
10:22:10,583  INFO || - [PartitionStateManager] hz.happy_perlman.client-query.thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Initializing cluster partition table arrangement...
10:22:10,621 DEBUG || - [JobRepository] hz.happy_perlman.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Job cleanup took 0ms
10:22:11,375  INFO |test_client_useJsonValue| - [SqlTestSupport] Time-limited test - Removing 0 cached plans in SqlTestSupport.@After
10:22:11,396  INFO |test_client_useJsonValue| - [SimpleTestInClusterSupport] Time-limited test - Ditching 0 jobs in SimpleTestInClusterSupport.@After: []
10:22:11,398  INFO |test_client_useJsonValue| - [SimpleTestInClusterSupport] Time-limited test - Destroying 2 distributed objects in SimpleTestInClusterSupport.@After: [hz:impl:replicatedMapService/__sql.catalog, hz:impl:mapService/test]
BuildInfo right after test_client_useJsonValue(com.hazelcast.jet.sql.SqlJsonTypeTest): BuildInfo{version='5.1-SNAPSHOT', build='20220104', buildNumber=20220104, revision=725d9c5, enterprise=false, serializationVersion=1}

No metrics recorded during the test
10:22:11,401 ERROR || - [EventServiceImpl] hz.happy_perlman.event-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] hz.happy_perlman.event-1 caught an exception while processing:com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher@3823397c
java.lang.UnsupportedOperationException: SQL catalog must never be cleared - MapEvent{entryEventType=CLEAR_ALL, member=Member [127.0.0.1]:5701 - b87c91fa-10ad-49a3-8c23-2840f726f150 this, name='__sql.catalog', numberOfEntriesAffected=1}
	at com.hazelcast.jet.sql.impl.schema.TablesStorage$EntryListenerAdapter.mapCleared(TablesStorage.java:186) ~[classes/:?]
	at com.hazelcast.replicatedmap.impl.ReplicatedMapEventPublishingService.dispatchEvent(ReplicatedMapEventPublishingService.java:132) ~[classes/:?]
	at com.hazelcast.replicatedmap.impl.ReplicatedMapService.dispatchEvent(ReplicatedMapService.java:232) ~[classes/:?]
	at com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher.run(LocalEventDispatcher.java:64) ~[classes/:?]
	at com.hazelcast.internal.util.executor.StripedExecutor$Worker.process(StripedExecutor.java:245) [classes/:?]
	at com.hazelcast.internal.util.executor.StripedExecutor$Worker.run(StripedExecutor.java:228) [classes/:?]

com.hazelcast.sql.HazelcastSqlException: java.lang.AssertionError: No assign rules for OTHER defined

	at com.hazelcast.sql.impl.QueryUtils.toPublicException(QueryUtils.java:74)
	at com.hazelcast.sql.impl.client.SqlClientService.rethrow(SqlClientService.java:297)
	at com.hazelcast.sql.impl.client.SqlClientService.rethrow(SqlClientService.java:288)
	at com.hazelcast.sql.impl.client.SqlClientService.execute(SqlClientService.java:113)
	at com.hazelcast.sql.SqlService.execute(SqlService.java:89)
	at com.hazelcast.jet.sql.SqlJsonTypeTest.executeClient(SqlJsonTypeTest.java:108)
	at com.hazelcast.jet.sql.SqlJsonTypeTest.test_client_useJsonValue(SqlJsonTypeTest.java:71)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: java.lang.AssertionError: No assign rules for OTHER defined
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:121)
	at com.hazelcast.client.impl.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:95)
	at com.hazelcast.client.impl.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:40)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:617)
	at com.hazelcast.sql.impl.client.SqlClientService.invoke(SqlClientService.java:263)
	at com.hazelcast.sql.impl.client.SqlClientService.execute(SqlClientService.java:109)
	... 15 more
Caused by: java.lang.AssertionError: No assign rules for OTHER defined
	at org.apache.calcite.sql.type.SqlTypeMappingRule.canApplyFrom(SqlTypeMappingRule.java:57)
	at org.apache.calcite.sql.type.SqlTypeUtil.canCastFrom(SqlTypeUtil.java:903)
	at org.apache.calcite.sql.type.SqlTypeFactoryImpl.leastRestrictiveByCast(SqlTypeFactoryImpl.java:184)
	at org.apache.calcite.sql.type.SqlTypeFactoryImpl.leastRestrictive(SqlTypeFactoryImpl.java:164)
	at com.hazelcast.jet.sql.impl.validate.types.HazelcastTypeFactory.leastRestrictive(HazelcastTypeFactory.java:173)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateValues(SqlValidatorImpl.java:5199)
	at org.apache.calcite.sql.validate.TableConstructorNamespace.validateImpl(TableConstructorNamespace.java:61)
	at org.apache.calcite.sql.validate.AbstractNamespace.validate(AbstractNamespace.java:89)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateNamespace(SqlValidatorImpl.java:1100)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateQuery(SqlValidatorImpl.java:1071)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validateQuery(HazelcastSqlValidator.java:184)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateInsert(SqlValidatorImpl.java:4589)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validateInsert(HazelcastSqlValidator.java:378)
	at org.apache.calcite.sql.SqlInsert.validate(SqlInsert.java:166)
	at com.hazelcast.jet.sql.impl.parse.SqlExtendedInsert.validate(SqlExtendedInsert.java:123)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validateScopedExpression(SqlValidatorImpl.java:1046)
	at org.apache.calcite.sql.validate.SqlValidatorImpl.validate(SqlValidatorImpl.java:752)
	at com.hazelcast.jet.sql.impl.validate.HazelcastSqlValidator.validate(HazelcastSqlValidator.java:179)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse0(QueryParser.java:80)
	at com.hazelcast.jet.sql.impl.parse.QueryParser.parse(QueryParser.java:57)
	at com.hazelcast.jet.sql.impl.OptimizerContext.parse(OptimizerContext.java:137)
	at com.hazelcast.jet.sql.impl.CalciteSqlOptimizer.prepare(CalciteSqlOptimizer.java:233)
	at com.hazelcast.sql.impl.SqlServiceImpl.prepare(SqlServiceImpl.java:261)
	at com.hazelcast.sql.impl.SqlServiceImpl.query0(SqlServiceImpl.java:241)
	at com.hazelcast.sql.impl.SqlServiceImpl.execute(SqlServiceImpl.java:198)
	at com.hazelcast.sql.impl.client.SqlExecuteMessageTask.call(SqlExecuteMessageTask.java:62)
	at com.hazelcast.client.impl.protocol.task.AbstractCallableMessageTask.processMessage(AbstractCallableMessageTask.java:35)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:153)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:116)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)
	at ------ submitted from ------.()
	at com.hazelcast.internal.util.ExceptionUtil.cloneExceptionWithFixedAsyncStackTrace(ExceptionUtil.java:279)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.returnOrThrowWithGetConventions(InvocationFuture.java:112)
	... 20 more

Finished Running Test: test_client_useJsonValue in 3.350 seconds.
10:22:11,441  INFO || - [SimpleTestInClusterSupport] Time-limited test - Terminating instance factory in SimpleTestInClusterSupport.@AfterClass
10:22:11,442  INFO || - [LifecycleService] Thread-2 - hz.client_1 [dev] [5.1-SNAPSHOT] HazelcastClient 5.1-SNAPSHOT (20220104 - 725d9c5) is SHUTTING_DOWN
10:22:11,450  WARN || - [TestClientRegistry$MockedServerConnection] pool-3-thread-1 - Server connection closed: null
10:22:11,451  INFO || - [MockServer] pool-3-thread-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:40001, connection: MockedNodeConnection{ remoteAddress = [127.0.0.1]:40001, localAddress = [127.0.0.1]:5701, connectionId = 1}
10:22:11,452  INFO || - [ClientConnectionManager] Thread-2 - hz.client_1 [dev] [5.1-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5701:b87c91fa-10ad-49a3-8c23-2840f726f150, connection: MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=false, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-01-04 10:22:11.351, lastWriteTime=2022-01-04 10:22:11.151, closedTime=2022-01-04 10:22:11.446, connected server version=5.1-SNAPSHOT}}
10:22:11,452  INFO || - [LifecycleService] Thread-2 - hz.client_1 [dev] [5.1-SNAPSHOT] HazelcastClient 5.1-SNAPSHOT (20220104 - 725d9c5) is CLIENT_DISCONNECTED
10:22:11,457  INFO || - [ClientEndpointManager] hz.happy_perlman.event-1 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Destroying ClientEndpoint{connection=MockedNodeConnection{ remoteAddress = [127.0.0.1]:40001, localAddress = [127.0.0.1]:5701, connectionId = 1}, clientUuid=fd4d94fe-7b69-47d7-a12d-6dcfe7adfd7f, clientName=hz.client_1, authenticated=true, clientVersion=5.1-SNAPSHOT, creationTime=1641280927927, latest clientAttributes=lastStatisticsCollectionTime=1641280928071,enterprise=false,clientType=JVM,clientVersion=5.1-SNAPSHOT,clusterConnectionTimestamp=1641280927894,clientAddress=127.0.0.1,clientName=hz.client_1,credentials.principal=null,os.committedVirtualMemorySize=7620087808,os.freePhysicalMemorySize=2693644288,os.freeSwapSpaceSize=16000217088,os.maxFileDescriptorCount=1048576,os.openFileDescriptorCount=300,os.processCpuTime=20370000000,os.systemLoadAverage=7.03,os.totalPhysicalMemorySize=33279762432,os.totalSwapSpaceSize=16000217088,runtime.availableProcessors=16,runtime.freeMemory=118247176,runtime.maxMemory=954728448,runtime.totalMemory=230162432,runtime.uptime=7615,runtime.usedMemory=111915256, labels=[]}
10:22:11,458  WARN || - [TwoWayBlockableExecutor] pool-3-thread-1 - Dropping incoming runnable since other end closed. Server Closed EOF. MockedClientConnection{localAddress=[127.0.0.1]:40001, super=ClientConnection{alive=false, connectionId=1, channel=null, remoteAddress=[127.0.0.1]:5701, lastReadTime=2022-01-04 10:22:11.351, lastWriteTime=2022-01-04 10:22:11.151, closedTime=2022-01-04 10:22:11.446, connected server version=5.1-SNAPSHOT}}
10:22:11,461  INFO || - [LifecycleService] Thread-2 - hz.client_1 [dev] [5.1-SNAPSHOT] HazelcastClient 5.1-SNAPSHOT (20220104 - 725d9c5) is SHUTDOWN
10:22:11,462  INFO || - [LifecycleService] Thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
10:22:11,469  WARN || - [Node] Thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Terminating forcefully...
10:22:11,469  INFO || - [Node] Thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Shutting down connection manager...
10:22:11,471  INFO || - [Node] Thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Shutting down node engine...
10:22:11,491  INFO || - [NodeExtension] Thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Destroying node NodeExtension.
10:22:11,492  INFO || - [Node] Thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] Hazelcast Shutdown is completed in 23 ms.
10:22:11,493  INFO || - [LifecycleService] Thread-2 - [127.0.0.1]:5701 [dev] [5.1-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN

Process finished with exit code 255
```

--- a/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastTypeFactory.java
+++ b/hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/validate/types/HazelcastTypeFactory.java
@@ -30,6 +30,7 @@ import static com.hazelcast.jet.sql.impl.validate.types.HazelcastTypeSystem.MAX_
 import static org.apache.calcite.sql.type.SqlTypeName.DECIMAL;
 import static org.apache.calcite.sql.type.SqlTypeName.DOUBLE;
 import static org.apache.calcite.sql.type.SqlTypeName.REAL;
+import static org.apache.calcite.sql.type.SqlTypeName.VARCHAR;
 
 /**
  * Custom Hazelcast type factory.
@@ -168,6 +169,31 @@ public final class HazelcastTypeFactory extends SqlTypeFactoryImpl {
 
     @Override
     public RelDataType leastRestrictive(List<RelDataType> types) {
+        // special-case for JSON - see https://github.com/hazelcast/hazelcast/issues/20303
+        // SqlTypeName for JSON is OTHER, there's missing handling for OTHER in SqlTypeAssignmentRule,
+        // and we don't know how to add it there. And even if we did, OTHER can represent both JSON and
+        // a JAVA object, and these aren't assignable.
+        boolean containsNullable = false;
+        boolean allJson = true;
+        boolean allJsonOrVarchar = true;
+        for (RelDataType type : types) {
+            if (!(type instanceof HazelcastJsonType)) {
+                allJson = false;
+            }
+            if (!(type instanceof HazelcastJsonType) && type.getSqlTypeName() != VARCHAR) {
+                allJsonOrVarchar = false;
+            }
+            if (type.isNullable()) {
+                containsNullable = true;
+            }
+        }
+        if (allJson) {
+            return containsNullable ? HazelcastJsonType.TYPE_NULLABLE : HazelcastJsonType.TYPE;
+        }
+        if (allJsonOrVarchar) {
+            return createSqlType(VARCHAR, containsNullable);
+        }
+
         // Calcite returns BIGINT for all integer types and DOUBLE for all inexact fractional types.
         // This code allows us to use more narrow types in these cases.
         RelDataType selected = super.leastRestrictive(types);

--- a/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlJsonTypeTest.java
+++ b/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/SqlJsonTypeTest.java
@@ -65,21 +65,34 @@ public class SqlJsonTypeTest extends SqlJsonTestSupport {
     public void test_client_useJsonValue() {
         createMapping(client(), "test", "bigint", "json");
 
-        final IMap<Long, HazelcastJsonValue> test = client().getMap("test");
-        test.put(1L, json("[1,2,3]"));
-
-        executeClient("INSERT INTO test VALUES (2, CAST('[4,5,6]' AS JSON))");
+        executeClient("INSERT INTO test VALUES (1, CAST('[4,5,6]' AS JSON))");
         assertRowsAnyOrder(client(),
-                "SELECT * FROM test" ,
-                rows(2, 1L, json("[1,2,3]"), 2L, json("[4,5,6]")));
+                "SELECT * FROM test WHERE __key=1" ,
+                rows(2, 1L, json("[4,5,6]")));
 
-        executeClient("INSERT INTO test VALUES (3, ?)", json("[7,8,9]"));
+        executeClient("INSERT INTO test VALUES (2, ?)", json("[7,8,9]"));
         assertRowsAnyOrder(client(),
-                "SELECT * FROM test" ,
+                "SELECT * FROM test WHERE __key=2",
                 rows(2,
-                        1L, json("[1,2,3]"),
-                        2L, json("[4,5,6]"),
-                        3L, json("[7,8,9]")
+                        2L, json("[7,8,9]")
+                ));
+
+        // both arguments JSON - test for https://github.com/hazelcast/hazelcast/issues/20303
+        executeClient("INSERT INTO test VALUES (3, CAST('[10,11,12]' AS JSON)), (4, ?)", json("[13,14,15]"));
+        assertRowsAnyOrder(client(),
+                "SELECT * FROM test WHERE __key>=3",
+                rows(2,
+                        3L, json("[10,11,12]"),
+                        4L, json("[13,14,15]")
+                ));
+
+        // one argument VARCHAR - test for https://github.com/hazelcast/hazelcast/issues/20303
+        executeClient("INSERT INTO test VALUES (5, '[16,17,18]'), (6, ?)", json("[19,20,21]"));
+        assertRowsAnyOrder(client(),
+                "SELECT * FROM test WHERE __key>=5",
+                rows(2,
+                        5L, json("[16,17,18]"),
+                        6L, json("[19,20,21]")
                 ));
     }
 

